File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        nodesSeen = set()
        curr = head

        while curr:
            if curr in nodesSeen:
                return True
            else:
                nodesSeen.add(curr)
            curr = curr.next

        return False

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // use two pointers
        // keep left pointer at first node,
        // move right pointers until r-l = n
        // then move both pointer at teh same time until r reaches the last node.
        // the Nth node from the end will be at the l + 1

        int l = 1;
        int r = 1;

        ListNode curr = head;
        while (r != n) {
            curr = curr.next;
            r++;
        }

        ListNode left = head;
        ListNode right = curr;
        ListNode prev = null;

        while (right.next != null) {
            prev = left;
            left = left.next;
            right = right.next;
        }

        ListNode rem = left; // removing this node

        if (rem == head) {
            ListNode temp = head.next;
            rem.next = null;
            return temp;
        }
        
        prev.next = rem.next;
        rem.next = null;

        return head;

    }
}
